Bug-fix related to large spectral flexure

tbowers7 opened the pull request:
This PR deals with the case of when the computed spectral flexure in a frame exceeds the FlexurePar `spec_maxshift` value.

In existing code, this situation yields a rather opaque `IndexError`, and it is not at all clear that the cause is excessive computed spectral flexure.  This PR clarifies the situation and adds a means for bypassing the error through a new FlexurePar Keyword `excessive_shift`, which can have one of three values:

- "crash" (default):  The equivalent behavior as the existing code, except that a more helpful `PypeItError` message specifying what happened is issued.
- "set_to_zero":  Skip the flexure correction for this object by setting the shift to zero, and moving on.
- "continue":  Apply the large correction while issuing a Warning message that the flexure shift exceeded `spec_maxshift`.

A paragraph is added to the Flexure doc specifying the new parameter and its behavior.

Unit tests pass:
```
$ pytest -Wignore
============================= test session starts ==============================
platform linux -- Python 3.9.12, pytest-7.1.2, pluggy-1.0.0
...
collected 200 items
...
======================= 200 passed in 228.16s (0:03:48) ========================

```
A #cloud run of the `dev-suite` will be required, but only after review of the PR has been completed.

tbowers7 committed:
Graceful catching of error when spectral shift > spec_maxshift

	modified:   doc/flexure.rst
	modified:   pypeit/core/flexure.py
	modified:   pypeit/extraction.py
	modified:   pypeit/par/pypeitpar.py

tbowers7 committed:
Modify error message.

	modified:   pypeit/core/flexure.py

debora-pe reviewed (approved):
This is very helpful @tbowers7. Thank you for doing it.
It looks good to me. I'm approving now, but please run the dev suite and if you can post some examples of the three cases.

tbowers7 committed:
Fixing typos

	modified:   pypeit/core/flexure.py

tbowers7 commented:
With existing code (`develop` branch), the situation of excessive measured shift yields an opaque downstream `IndexError` in `pypeit/extraction.py`:
```
[WARNING] :: flexure.py 136 spec_flex_shift() - If we use Paranal, cut down on wavelength early on
[INFO]    :: flexure.py 164 spec_flex_shift() - Resolution of Archive=38468.442229331384 and Observation=235.78259307812402
[WORK IN ]::
[PROGRESS]:: flexure.py 239 spec_flex_shift() - Need to mask bad pixels
[WORK IN ]::
[PROGRESS]:: flexure.py 242 spec_flex_shift() - Consider taking median first [5 pixel]
[WORK IN ]::
[PROGRESS]:: flexure.py 257 spec_flex_shift() - Consider taking median first [5 pixel]
[WARNING] :: flexure.py 278 spec_flex_shift() - Flexure compensation failed for one of your objects
[WARNING] :: flexure.py 426 spec_flexure_slit() - Flexure shift calculation failed for this spectrum.
[INFO]    :: flexure.py 453 spec_flexure_slit() - No flexure corrections could be made
Traceback (most recent call last):
  File "/Users/tbowers/opt/miniconda3/envs/pypeit/bin/run_pypeit", line 33, in <module>
    sys.exit(load_entry_point('pypeit', 'console_scripts', 'run_pypeit')())
  File "/Users/tbowers/d1/codes/PypeIt/pypeit/scripts/scriptbase.py", line 110, in entry_point
    cls.main(cls.parse_args())
  File "/Users/tbowers/d1/codes/PypeIt/pypeit/scripts/run_pypeit.py", line 142, in main
    pypeIt.reduce_all()
  File "/Users/tbowers/d1/codes/PypeIt/pypeit/pypeit.py", line 447, in reduce_all
    sci_spec2d, sci_sobjs = self.reduce_exposure(frames, bg_frames=bg_frames,
  File "/Users/tbowers/d1/codes/PypeIt/pypeit/pypeit.py", line 620, in reduce_exposure
    = self.extract_one(frames, self.det, sciImg_list[i], objFind_list[i],
  File "/Users/tbowers/d1/codes/PypeIt/pypeit/pypeit.py", line 860, in extract_one
    tilts = self.exTract.run(final_global_sky, ra=self.fitstbl["ra"][frames[0]],
  File "/Users/tbowers/d1/codes/PypeIt/pypeit/extraction.py", line 437, in run
    self.spec_flexure_correct(mode='local', sobjs=self.sobjs)
  File "/Users/tbowers/d1/codes/PypeIt/pypeit/extraction.py", line 559, in spec_flexure_correct
    new_sky = sobj.apply_spectral_flexure(this_flex_dict['shift'][cntr],
IndexError: list index out of range
```

--------
Example of the new (default) `crash` option for `excessive_shift` with the same object as above (same functional effect as the existing code, but with a useful error message):
```
[WARNING] :: flexure.py 141 spec_flex_shift() - If we use Paranal, cut down on wavelength early on
[INFO]    :: flexure.py 169 spec_flex_shift() - Resolution of Archive=38468.442229331384 and Observation=195.31620707028867
[WORK IN ]::
[PROGRESS]:: flexure.py 244 spec_flex_shift() - Need to mask bad pixels
[WORK IN ]::
[PROGRESS]:: flexure.py 247 spec_flex_shift() - Consider taking median first [5 pixel]
[WORK IN ]::
[PROGRESS]:: flexure.py 262 spec_flex_shift() - Consider taking median first [5 pixel]
[WARNING] :: flexure.py 284 spec_flex_shift() - Computed shift 29.2 pix is larger than specified maximum 20 pix.

[ERROR]   :: flexure.py 288 spec_flex_shift() - Flexure compensation failed for one of your
             objects.  Either adjust the "spec_maxshift"
             FlexurePar Keyword, or see the flexure documentation
             for information on how to bypass this error using the
             "excessive_shift" keyword.
             https://pypeit.readthedocs.io/en/release/flexure.html
Traceback (most recent call last):
...
pypeit.pypmsgs.PypeItError: Flexure compensation failed for one of your
             objects.  Either adjust the "spec_maxshift"
             FlexurePar Keyword, or see the flexure documentation
             for information on how to bypass this error using the
             "excessive_shift" keyword.
             https://pypeit.readthedocs.io/en/release/flexure.html
```

--------
Example of the `set_to_zero` option for `excessive_shift`, which skips flexure correction by setting the shift to zero:
```
[WARNING] :: flexure.py 142 spec_flex_shift() - If we use Paranal, cut down on wavelength early on
[INFO]    :: flexure.py 170 spec_flex_shift() - Resolution of Archive=38468.442229331384 and Observation=195.31620707028867
[WORK IN ]::
[PROGRESS]:: flexure.py 245 spec_flex_shift() - Need to mask bad pixels
[WORK IN ]::
[PROGRESS]:: flexure.py 248 spec_flex_shift() - Consider taking median first [5 pixel]
[WORK IN ]::
[PROGRESS]:: flexure.py 263 spec_flex_shift() - Consider taking median first [5 pixel]
[WARNING] :: flexure.py 285 spec_flex_shift() - Computed shift 29.2 pix is larger than specified maximum 20 pix.
[WARNING] :: flexure.py 297 spec_flex_shift() - Flexure compensation failed for one of your objects.
[WARNING] :: flexure.py 298 spec_flex_shift() - Setting the flexure correction shift to 0 pixels.
[INFO]    :: flexure.py 323 spec_flex_shift() - Flexure correction of 0.000 pixels

```

--------
Finally, example of the `continue` option, which accepts the excessively large shift and moves on:
```
[WARNING] :: flexure.py 141 spec_flex_shift() - If we use Paranal, cut down on wavelength early on
[INFO]    :: flexure.py 169 spec_flex_shift() - Resolution of Archive=38468.442229331384 and Observation=195.31620707028867
[WORK IN ]::
[PROGRESS]:: flexure.py 244 spec_flex_shift() - Need to mask bad pixels
[WORK IN ]::
[PROGRESS]:: flexure.py 247 spec_flex_shift() - Consider taking median first [5 pixel]
[WORK IN ]::
[PROGRESS]:: flexure.py 262 spec_flex_shift() - Consider taking median first [5 pixel]
[WARNING] :: flexure.py 284 spec_flex_shift() - Computed shift 29.2 pix is larger than specified maximum 20 pix.
[WARNING] :: flexure.py 305 spec_flex_shift() - Applying flexure shift larger than specified max!
[INFO]    :: flexure.py 320 spec_flex_shift() - Flexure correction of 29.181 pixels
```

tbowers7 committed:
Change meaning of `excessive_shift = skip` parameter

Setting `excessive_shift` to `skip` now skips 1D spectral extraction, as if
there were no objects found in the frame.

This change was made because it is unlikely that a shift of zero pixels is
correct, and simply skipping extraction has the effect of removing the
problematic frame from the final list of 1D spectra.

	modified:   doc/flexure.rst
	modified:   doc/pypeit_par.rst
	modified:   pypeit/core/flexure.py
	modified:   pypeit/extraction.py
	modified:   pypeit/par/pypeitpar.py

tbowers7 commented:
Updated the meaning of `excessive_shift = skip` to skip the 1D spectral extraction as if no objects had been found in the first place.  The example above has been modified to reflect the new meaning.  @deborape , please review the updated PR.

**NOTE:** This change has been reverted (2-Aug)

debora-pe commented on pypeit/extraction.py:
```diff
@@ -391,25 +392,29 @@ def run(self, global_sky, prepare_extraction=True, model_noise=None, spat_pix=No
+        #   specified shift will be NaN.  Skip object extraction by setting
+        #   self.sobjs_obj to a blank container (i.e., self.nobj_to_extract = 0)
+        if np.logical_not(np.isfinite(self.slitshift)).any():
+            self.sobjs_obj = specobjs.SpecObjs()
```
I don't think it's necessary to skip the extraction of all the objects. Especially for multislit spectroscopy it's more common to have the flexure correction to fail for only one or a few slits. So, instead of setting `self.sobjs_objs` to a black container, we can just remove the object (using the method `remove_sobj`, there are some examples above) that is in the slit with bad/failed flexure correction. 

I also want to mention that below a local flexure correction is applied (i.e., the global correction is fine tuned for each specific extracted object) and it may happen that the flexure correction fails there too.

tbowers7 replied:
The scope of changes needed for skipping extraction is getting beyond my level of expertise.  I have therefore reverted this last change (there is no more `skip_extraction` option) and renamed `skip` --> `set_to_zero`.

The machinery is in place now for someone more skilled to add a `skip_extraction` option for `excessive_shift` if desired.

debora-pe commented on pypeit/core/flexure.py:
```diff
@@ -459,24 +459,20 @@ def spec_flexure_slit(slits, slitord, slit_bpm, sky_file, method="boxcar", speco
+                        # One does not exist yet -- break out of the FOR loop
+                        break
+                        # # Save it for later
+                        # return_later_sobjs.append([islit, ss])
```
This basically skips the code at lines 481-499 and `return_later_sobjs` is never used. It's probably better to comment that part of the code and add a note. I would also check with the person that wrote this part of code (maybe @profxj) if it's OK with it.

tbowers7 replied:
I believe this change was purely cosmetic (removing the `punt` business and putting the `break` where `punt = True` was set), but I have reverted it alongside the other changes as mentioned above.

debora-pe commented on pypeit/core/flexure.py:
```diff
@@ -294,12 +294,9 @@ def spec_flex_shift(obj_skyspec, arx_skyspec, arx_lines, mxshft=20, excess_shft=
-                            corr=corr[subpix_grid.astype(np.int)],
-                            sky_spec=obj_skyspec, arx_spec=arx_skyspec,
-                            corr_cen=corr.size/2, smooth=smooth_sig_pix)
+                msgs.warn("Skipping 1D spectral extraction.")
```
I appreciate the possibility to skip the extraction of the 1D extraction if the flexure correction fails, but I thought that the option to set it to zero was also a good one. Can we have both `skip_extraction` and `set_to_zero` options for `excess_shft`?

tbowers7 replied:
Have reverted last change, and changed `skip` to `set_to_zero`.

debora-pe reviewed (changes requested):
@tbowers7 Thanks for showing some examples and adding these changes. 

I'm requesting some modifications. See my comments.

tbowers7 committed:
Revert `skip_extraction`; change `skip` --> `set_to_zero`

Reverting all `skip_extraction` changes from previous commit, as the details of
implementation are beyond my expertise level.

Change the name of the `skip` option to `set_to_zero` to more clearly indicate
the effect of use.

	modified:   doc/flexure.rst
	modified:   doc/pypeit_par.rst
	modified:   pypeit/core/flexure.py
	modified:   pypeit/extraction.py
	modified:   pypeit/par/pypeitpar.py

tbowers7 committed:
Merge branch 'develop' into excess_flexure

debora-pe reviewed (approved):
@tbowers7 sounds good to me!

badpandabear commented:
Dev Suite results:
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  200 passed, 1195 warnings in 199.29s (0:03:19) ---
--- PYTEST UNIT TESTS PASSED  115 passed, 1803 warnings in 1090.25s (0:18:10) ---
--- PYTEST VET TESTS PASSED  24 passed, 1566 warnings in 969.10s (0:16:09) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 1/152 TESTS  ---
Failed tests:
    ldt_deveny/DV8 pypeit 
Skipped tests:
Testing Started at 2022-08-03T16:46:16.028044
Testing Completed at 2022-08-04T00:55:00.673774
Total Time: 8:08:44.645730

```
The one failure was:

```
-------------------------
Test Setup: ldt_deveny/DV8

-------------------------
Directories:
         Raw data: /PypeIt-development-suite/RAW_DATA/ldt_deveny/DV8
    PypeIt output: /tmp/REDUX_OUT/ldt_deveny/DV8
Files:
     .pypeit file: None
 Std .pypeit file: None
Tests:
----
ldt_deveny/DV8 pypeit  Result: --- FAILED

Logfile:    /tmp/REDUX_OUT/ldt_deveny/DV8/ldt_deveny_dv8.test.log
Process Id: 2565
Start time: Wed Aug  3 23:25:52 2022
End time:   Wed Aug  3 23:30:17 2022
Duration:   0:04:24.582998
Command:    run_pypeit /tmp/REDUX_OUT/ldt_deveny/DV8/ldt_deveny_dv8.pypeit -o

Error Messages:

End of Log:
             for information on how to bypass this error using the
             "excessive_shift" keyword.
             https://pypeit.readthedocs.io/en/release/flexure.html

```

jhennawi commented on doc/flexure.rst:
```diff
@@ -106,6 +106,18 @@ are respectively used (see `Alternate sky models`_ for all sky models).
+execution with an error.  Optionally, the user may sepecify ``excessive_shift = set_to_zero``
+to allow the code to continue executing while skipping flexure correction (for this object)
+by setting the shift to zero or ``excessive_shift = continue`` to utilize the large shift value.
+This festure is included *caveat emptor*, and the user should carefully examine the
```
feature --> feature

tbowers7 replied:
Thanks for finding the typo

jhennawi reviewed (approved):
Thanks for this @tbowers7. It is great to see our cryptic crash messages made more clear and the improved documentation.

tbowers7 committed:
Fixed typo

	modified:   doc/flexure.rst

tbowers7 committed:
Merge branch 'develop' into excess_flexure

tbowers7 commented:
Waiting on resolution of the excessive flexure fail in DV8 that did not occur on my own machine (DevSuite `master`, PypeIt `excess_flexure`) but did occur in the #cloud.
```
$ ./pypeit_test reduce -i ldt_deveny -s DV8
Running reduce tests.
Running tests on the following instruments:
    ldt_deveny

Reducing data from ldt_deveny for the following setups:
    DV8

 0 passed/ 0 failed/ 0 skipped STARTED ldt_deveny/DV8 pypeit 
 1 passed/ 0 failed/ 0 skipped PASSED  ldt_deveny/DV8 pypeit 
--- PYPEIT DEVELOPMENT SUITE PASSED 1/1 TESTS  ---

Test Summary
--------------------------------------------------------
--- PYPEIT DEVELOPMENT SUITE PASSED 1/1 TESTS  ---
Testing Started at 2022-08-08T10:21:16.346325
Testing Completed at 2022-08-08T10:39:09.718339
Total Time: 0:17:53.372014
```
Inspection of the `spec1d` FITS files output from the DevSuite run showed flexure shifts all < 1 pixel, well below the 20-pixel limit specified by `spec_maxshift`.

tbowers7 commented:
From @badpandabear :

> Running that setup individually in the cloud has passed twice now. It's also passed twice when I ran it locally. Which is annoying because I don't know what happened the first time.  I'm going to try a full run of the dev-suite for my own curiosity but I don't  know that you need to wait for that.
